wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1425,3 +1392,288 @@ let construct_pointer_constraint (l_orig : Constr.z3_expr list) (env1 : Env.t)
+  let find_depth depth node = Env.Unroll_depth.find depth (Node.label node)
+                              |> Option.value ~default:num_unroll
+  in
+  let decr_depth depth node =
+    Env.Unroll_depth.update depth (Node.label node)
+      ~f:(function None -> num_unroll - 1 | Some n -> n - 1)
+  in
+  let unroll env pre ~start:node g =
+    let tid = node |> Node.label |> Term.tid in
+    debug "Unrolling loop for %s%!" (Tid.to_string tid);
+    let depth = Env.get_unroll_depth env in
+    if find_depth depth node <= 0 then
+      let pre =
+        List.find_map [loop_exit] ~f:(fun spec ->
+            let* exit = spec node g in
+            let tid = exit |> Node.label |> Term.tid in
+            Env.get_precondition env tid)
+      in
+      let pre = match pre with
+        | Some p -> p
+        | None ->
+          warning "Trivial precondition is being used for node %s%!" (Tid.to_string tid);
+          let ctx = Env.get_context env in
+          Constr.trivial ctx
+      in
+      Env.add_precond env tid pre
+    else
+      begin
+        let updated_depth = decr_depth depth node in
+        let updated_env = Env.set_unroll_depth env updated_depth in
+        let _, env = visit_graph updated_env pre
+            ~start:node ~skip_node:(unreachable_from_start g node) g in
+        env
+      end
+  in
+  unroll
+
+(* Gets the target destination of a jump. We do not handle indirect jumps. *)
+let target (jmp : Jmp.t) : Tid.t option =
+  match Jmp.kind jmp with
+  | Goto (Direct tid) -> Some tid
+  | Int (_, tid) -> Some tid
+  | Call t -> begin
+      match Call.return t with
+      | Some (Direct tid) -> Some tid
+      | Some (Indirect _) | None -> None
+    end
+  | Goto (Indirect _) | Ret _ -> None
+
+(* Checks if [tid] corresponds to the tid of a block in a loop. *)
+let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
+  Seq.exists (Group.enum loop) ~f:(fun n ->
+      Tid.equal (n |> Graphs.Ir.Node.label |> Term.tid) tid)
+
+(* Assumes the loop header branches into two blocks. *)
+let get_branch (blk : Blk.t) : (Jmp.t * Jmp.t) option =
+  match Seq.to_list (Term.enum jmp_t blk) with
+  | [jmp1; jmp2] -> Some (jmp1, jmp2)
+  | _ -> None
+
+(* Gets the BAP expression that represents the condition for going into the
+   loop. *)
+let get_branch_condition (loop : Graphs.Ir.Node.t group) (jmp1 : Jmp.t)
+    (jmp2 : Jmp.t) : Exp.t option =
+  match target jmp1, target jmp2 with
+  | Some target1, Some target2 ->
+    if in_loop loop target1 then
+      (* jmp1 goes to loop, jmp2 exits. while(E) is jmp1's condition. *)
+      Some (Jmp.cond jmp1)
+    else if in_loop loop target2 then
+      (* jmp1 exits, jmp2 goes to loop. while(E) is the negation of jmp1's
+         condition. *)
+      Some Bil.(unop not (Jmp.cond jmp1))
+    else
+      None
+  | _, _ -> None
+
+(* Gets a constraint representing the condition for a loop. *)
+let get_cond (loop : Graphs.Ir.Node.t group) (blk : Blk.t) (env : Env.t)
```

Comment:
Again, this feels like code duplication from `visit_blk`. I wonder if we can generalize it to handle both cases.

Reply:
The `compute` in `get_cond` is basically the same as `visit_blk` except that it skips jumps, so I can probably do something similar to what I did with `visit_graph` and pass in a function that determines which instructions to skip.

Reply:
Might be the right thing. Let's discuss maybe.